src/terminal.rs

```diff
@@ -471,38 +471,48 @@ where
             return Ok(());
         }
 
+        // Clear the viewport off the screen
         self.clear()?;
-        let height = height.min(self.last_known_size.height);
-        self.backend.append_lines(height)?;
-        let missing_lines =
-            height.saturating_sub(self.last_known_size.bottom() - self.viewport_area.top());
+
+        // Move the viewport accordingly - if it's not already at the bottom of the terminal
+        self.set_viewport_area(Rect {
+            x: self.viewport_area.left(),
+            y: (self.viewport_area.top() + height)
+                .min(self.last_known_size.bottom() - self.viewport_area.height),
+            width: self.viewport_area.width,
+            height: self.viewport_area.height,
+        });
+
+        // Draw contents into buffer
         let area = Rect {
             x: self.viewport_area.left(),
-            y: self.viewport_area.top().saturating_sub(missing_lines),
+            y: 0,
             width: self.viewport_area.width,
             height,
         };
         let mut buffer = Buffer::empty(area);
-
         draw_fn(&mut buffer);
 
-        let iter = buffer.content.iter().enumerate().map(|(i, c)| {
-            let (x, y) = buffer.pos_of(i);
-            (x, y, c)
-        });
-        self.backend.draw(iter)?;
-        self.backend.flush()?;
-
-        let remaining_lines = self.last_known_size.height - area.bottom();
-        let missing_lines = self.viewport_area.height.saturating_sub(remaining_lines);
-        self.backend.append_lines(self.viewport_area.height)?;
-
-        self.set_viewport_area(Rect {
-            x: area.left(),
-            y: area.bottom().saturating_sub(missing_lines),
-            width: area.width,
-            height: self.viewport_area.height,
-        });
+        // Split buffer into screen-sized chunks and draw
+        let max_chunk_size = (self.viewport_area.top() * area.width).into();
+        for buffer_content_chunk in buffer.content.chunks(max_chunk_size) {
+            let chunk_size = buffer_content_chunk.len() as u16 / area.width;
+
+            self.backend
+                .append_lines(self.viewport_area.height.saturating_sub(1) + chunk_size)?;
+
+            let iter = buffer_content_chunk.iter().enumerate().map(|(i, c)| {
+                let (x, y) = buffer.pos_of(i);
+                (
+                    x,
+                    self.viewport_area.top().saturating_sub(chunk_size) + y,
+                    c,
+                )
+            });
+            self.backend.draw(iter)?;
+            self.backend.flush()?;
+            self.set_cursor(self.viewport_area.left(), self.viewport_area.top())?;
+        }
```

Comment:
This is fairly complex, and I'm not sure I understand it correctly. It would be helpful to explain this with a bit more of a comment. I'm going to write out an example to help clarify my understanding (I'm trying to use numbers that don't accidentally overlap to make it clearer)

Say the screen size is 40 lines, the viewport is 25 lines high starting at line 5, and we want to add 20 more lines, then I think the effect of this method is to move the viewport to start at line 15, then we create a 20 line buffer and call the draw function. Then we divide the buffer into 2 chunks sized 15 and 5.

In the first loop iteration we append (25-1+15) = 39 lines (why?) and then iterate over the chunk and draw those lines into y=0-15.

In the second loop iteration we append (25-1+5) = 29 lines and then iterate over the chunk and draw those lines into y=0-5

I wonder if `self.backend.append_lines(self.viewport_area.height.saturating_sub(1) + chunk_size)?;` is meant to be `self.backend.append_lines(self.viewport_area.height.saturating_sub(1 + chunk_size))?;`

I wonder if perhaps to test this, we should use a backend trait implementation that records each operation (i.e. we're looking at how terminal and backed interact rather than the output).

Reply:
If the screen is 40 lines, a viewport of 25 starting at line 5 (index 4) and we wish to add 20 lines:

Firstly we move the viewport by height:

1. `viewport_at_bottom` is calculated as `terminal height - viewport height` 40 - 25 = 15
2. Move the viewport `(viewport.y + add 20 lines).min(viewport_at_bottom)` (4 + 20).min(15) = 15

In this case the viewport is now at the bottom of the screen.

Next, we calculate the `max_chunk_size`, this being the most we can draw to the screen at a time assuming we cannot draw over the viewport. I did this for a couple reasons;

- We can only draw to the screen what we can see, then `append_lines` will scroll this out of the view of the terminal, so this maximum accommodates for the cursor position (which is at the start of the viewport in this point in time) and allows us to insert `n > terminal height`.
- I was hoping there would be a way to remove the `self.clear()` in order to reduce the flickering that I've been experiencing at high throughput. This is yet to bear fruit. The concept however is that if newlines are only appended *before* the viewport then maybe there is less reason to clear the viewport.

Back to the example with `viewport_area.top()` now at 15:
max_chunk_size = 15 * width (or 15 lines)

Now conceptually the for loop scrolls the terminal by 15 and then inserts 15-line chunks of the buffer at a time.

#### Alternative approaches

I see that a possible alternative is perhaps to instead use the entire terminal screen as the `max_chunk_size` but I think this precludes us to potential solutions to removing the `self.clear()`, perhaps something along the lines of using the diffing capabilities instead, in the future.

#### Comment on testing

> I wonder if perhaps to test this, we should use a backend trait implementation that records each operation (i.e. we're looking at how terminal and backed interact rather than the output).

I think this would be a great deal of work to implement on a TestBackend of sorts. I could potentially see the benefit but personally I think the tests I've added to `test/terminal.rs` cover the cases I was initially concerned with.

I've done some manual testing along simular lines, I've tested with `time::sleeps` in a couple of configurations making sure the step-by-step behaviour makes sense and it does indeed fix the issues I had (minus the flickering).

---

I hope this explanation has provided some clarity on my thoughts about the process. Thanks for your feedback :)

Reply:
Ok - got it. I understand the main part of this now.

I think you're right that setting up TestBackend to test this would be difficult. I think a more manual approach will have to do for testing this on the real terminal for now.